File: week4/java/CombinationSum_39.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();
        combination(candidates, 0, target, new ArrayList<Integer>(), result);
        return result;
    }
    
    private void combination(int[] candidates, int index, int target, List<Integer> list, List<List<Integer>> result) {
        
        if (target < 0) return;
        
        if (target == 0) {
            result.add(new ArrayList<Integer>(list));
        }
        
        for (int i = index; i < candidates.length; i++) {
            int tmp = target - candidates[i];
            
            if (tmp < 0) continue;
            
            list.add(candidates[i]);
            
            if (tmp == 0) {
                result.add(new ArrayList<>(list));
            } else {
                combination(candidates, i, tmp, list, result);
            }
            list.remove(list.size() - 1);
        }
    }
}

File: week4/python/CombinationSum_39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        result = list()
        candidates = sorted(candidates, reverse=True)
        
        def combination(cur_list: List[int], cur_sum: int, index: int):
            if index < 0 or cur_sum > target:
                return
            
            if cur_sum == target:
                result.append(list(cur_list))
            
            for i in range(index, -1, -1):
                if cur_sum + candidates[i] > target:
                    break
                    
                cur_list.append(candidates[i])
                combination(cur_list, cur_sum + candidates[i], i)
                cur_list[:] = cur_list[:-1]
                
        combination(list(), 0, len(candidates) - 1)
        return result
